ICU-23053 In the ICU4X mode of the collation builder, mark middle starters in contractions

hsivonen opened the pull request:
This change uses a contraction ce32 for starters that occur in the middle of contractions but do not themselves start contractions. In practice, this affects two characters in the root (U+0CC2 and U+0DCF) and none in the CLDR tailorings.

This enables finding out whether a pair of starters might participate in a contraction without having to look further back, which should enable the identical prefix optimization directly from ce32s.

#### Checklist
- [x] Required: Issue filed: ICU-23053
- [x] Required: The PR title must be prefixed with a JIRA Issue number. Example: "ICU-1234 Fix xyz"
- [x] Required: Each commit message must be prefixed with a JIRA Issue number. Example: "ICU-1234 Fix xyz"
- [x] Issue accepted (done by Technical Committee after discussion)
- [ ] Tests included, if applicable
- [ ] API docs and/or User Guide docs changed or added, if applicable

I have tested this using the ICU4X tree (not landed yet).

sffc commented:
Hi, it seems that this blocks ICU4X 2.0 since it breaks data. Can the review be prioritized so that we can tag an icuexportdata release?

hsivonen commented:
Related: https://github.com/unicode-org/unicodetools/issues/1101

markusicu commented on icu4c/source/i18n/collationdatabuilder.cpp:
```diff
@@ -1391,7 +1419,71 @@ CollationDataBuilder::buildMappings(CollationData &data, UErrorCode &errorCode)
+        // matcher.
+        UnicodeString placeholder(0xDC00);
+
+        UnicodeSetIterator iter(contractionMiddleStarter);
+        while(iter.next() && !iter.isString()) {
```
optional but nice to have:
something like this should work since ICU 76 --
```
for (UChar32 c : contractionMiddleStarter.codePoints()) {
```

see https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/classicu_1_1UnicodeSet.html#a537a13a1b3adcaebc140f715876aca63

hsivonen replied:
Changed.

markusicu commented on icu4c/source/i18n/collationdatabuilder.cpp:
```diff
@@ -1391,7 +1419,71 @@ CollationDataBuilder::buildMappings(CollationData &data, UErrorCode &errorCode)
+                if (Collation::hasCE32Tag(ce32, Collation::CONTRACTION_TAG)) {
+                    // This middle starter is also the first character of another
+                    // contraction, but that contraction does not have the
+                    // CONTRACT_HAS_STARTER flag. Let's add the flag to make
+                    // mark this at the expense of pessimizing the matching
+                    // of this contraction.
```
```suggestion
                    // CONTRACT_HAS_STARTER flag. Let's add the flag to
                    // mark this at the expense of pessimizing the matching
                    // of this contraction.
```

markusicu commented on icu4c/source/i18n/collationdatabuilder.cpp:
```diff
@@ -1391,7 +1419,71 @@ CollationDataBuilder::buildMappings(CollationData &data, UErrorCode &errorCode)
+                    // This middle starter is not also the first character
+                    // in another contraction.
+
+                    // The UCharsTrie needs to contain some placeholder. Trying to
+                    // build an empty UCharsTrie causes an error. We build a trie
```
```suggestion
                    // The UCharsTrie needs to contain some placeholder
                    // because it cannot be empty. We build a trie
```

markusicu commented on icu4c/source/i18n/collationdatabuilder.cpp:
```diff
@@ -1391,7 +1419,71 @@ CollationDataBuilder::buildMappings(CollationData &data, UErrorCode &errorCode)
+                    // matcher and we put an unpaired low surrogate into the trie.
+                    // This pessimizes the character to CE mapping of the `c`,
+                    // since useless trie matching will be attempted but as of
+                    // February 2025, only two characters are affected.
```
```suggestion
                    // February 2025, only two relatively rare characters are affected.
```

markusicu commented on icu4c/source/i18n/collationdatabuilder.h:
```diff
@@ -254,6 +254,12 @@ class U_I18N_API CollationDataBuilder : public UObject {
     UnicodeSet unsafeBackwardSet;
+    /**
+     * For ICU4X only: The starters that occur in some contraction
+     * in a position that is neither the first nor the last scalar
```
```suggestion
     * in a position that is neither the first nor the last code point
```

markusicu commented on icu4c/source/i18n/collationdatabuilder.cpp:
```diff
@@ -655,6 +655,20 @@ CollationDataBuilder::addCE32(const UnicodeString &prefix, const UnicodeString &
                     return;
                 }
             }
+            int32_t sCount = s.countChar32();
+            UChar32 sUtf32[32];
+            int32_t sLen = s.toUTF32(sUtf32, 32, errorCode);
+            if (sLen != sCount) {
+                // If this error is ever reached, just increase the buffer
+                // size above.
+                errorCode = U_UNSUPPORTED_ERROR;
+                return;
+            }
```
FYI: no real need to convert to a UTF-32 string; you could consider some alternatives, or keep this as is

- UnicodeString lets you iterate over its code points using char32At() and moveIndex32(index, delta=1)
- ICU 78 adds a real C++ string code point iterator ([unicode/utfiterator.h](https://github.com/unicode-org/icu/blob/main/icu4c/source/common/unicode/utfiterator.h) function `utfIterator<UChar32, UTF_BEHAVIOR_SURROGATE>(s)` makes a “range”)
  - You could use `auto limit = range.end();` then `--limit;` to skip the last code point, skip `auto start = range.begin();` via `++start` if it's not already at the limit, and then explicitly loop `for (; start != limit; ++start)` and use `start->codePoint()`

FYI 2
I also feel like the builder probably doesn't allow contractions longer than 31 UChars but can't find a constant for that so maybe that's not currently a safe/enforced assumption

hsivonen replied:
I think I'll leave this as-is for now in case a need to backport to ICU4C earlier than 78 shows up.

markusicu commented:
Clever; I should try to do something like this in ICU4C/J if I ever find time to work on that again...

What do you do in identical-prefix matching with ccc!=0 characters? Additional lookup in normalization data?

markusicu commented on icu4c/source/i18n/collationdatabuilder.cpp:
```diff
@@ -1391,7 +1419,71 @@ CollationDataBuilder::buildMappings(CollationData &data, UErrorCode &errorCode)
+
+                    int32_t index = addContextTrie(ce32, contractionBuilder, errorCode);
+                    if(U_FAILURE(errorCode)) { return; }
+                    if(index > Collation::MAX_INDEX) {
```
Does addContextTrie() not already check this?

hsivonen replied:
If it is checked, it's not obvious. Pre-existing calls to the method in this file had this check right after, which is why I copied the check here.

markusicu reviewed (commented):
Mostly lgtm. Thanks for the comprehensive comments.
And sorry for the long delay.

hsivonen commented:
I believe I have now addressed the review comments.

markusicu commented:
> I believe I have now addressed the review comments.

Thanks, will look. In the future, please just add commits when responding to feedback, and squash only at the very end after approval.

markusicu commented:
@hsivonen are you ready for me to merge? Can I also close the ticket, or is there more to be done?

hsivonen commented:
> > I believe I have now addressed the review comments.
> 
> Thanks, will look. In the future, please just add commits when responding to feedback, and squash only at the very end after approval.

OK.

> @hsivonen are you ready for me to merge? Can I also close the ticket, or is there more to be done?

Yes, please merge. It's OK to close the ticket on the ICU4C side.

Thanks!
